core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/MemoryQueue.scala

```diff
@@ -154,8 +168,8 @@ class MemoryQueue(private val etcdClient: EtcdClient,
   private val staleQueueRemovedMsg = QueueRemoved(invocationNamespace, action.toDocId.asDocInfo(revision), None)
   private val actionRetentionTimeout = MemoryQueue.getRetentionTimeout(actionMetaData, queueConfig)
 
-  private[queue] var containers = Set.empty[String]
-  private[queue] var creationIds = Set.empty[String]
+  private[queue] var containers = java.util.concurrent.ConcurrentHashMap.newKeySet[String]().asScala
```

Comment:
This was one of the reasons for failed activations during the deployment.
The `Set` was not thread-safe.

Reply:
This was a concern for me in the past with all of the `vars` in MemoryQueue. Are we sure this covers all var cases now to be thread safe? 

I'm trying to debug an issue where etcd keys are orphaned for running containers (not warmed) that do not exist when I disable a scheduler and do a rolling deployment. Can reproduce it anytime I disable a scheduler right now. It's a new issue and think it may have been introduced with the cycle sending commit https://github.com/apache/openwhisk/commit/ef725a653ab112391f79c274d8e3dcfb915d59a3; I'm wondering if this actually fixes it.

Reply:
Hm.. first, the etcd keys for running containers are handled by invokers so it's surprising you face the issue when you do a rolling deployment of schedulers. I think you need to trace the problematic container in the invoker.

I haven't looked into it deeply yet but if there are any issues, it could be because of this though I haven't faced any so far.
Not sure if there are any easily replaceable alternatives.

```
private var requestBuffer = mutable.PriorityQueue.empty[BufferedRequest]
```

Others are basically immutable collections, so I think it would be ok to use them in the context of Akka actor.
If this really needs to be updated, I hope it is handled in a subsequent PR.